src/mtcp/mtcp_restart.c

```diff
@@ -632,19 +663,31 @@ main(int argc, char *argv[], char **environ)
     //   before vvar and after vdso, and verify that we get an EFAULT.
     // In May, 2020, on Cori and elsewhere, vvar is 3 pages and vdso is 2 pages.
     char *vdsoStart = (char *)mygetauxval(environ, AT_SYSINFO_EHDR);
-    char *vvarStart = vdsoStart - 3 * PAGESIZE;
-    void *rc1 = mtcp_sys_mremap(vvarStart, 3*PAGESIZE, 3*PAGESIZE,
-		                MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
-    void *rc2 = mtcp_sys_mremap(vdsoStart, 2*PAGESIZE, 2*PAGESIZE,
-		                MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
-    if (rc2 == MAP_FAILED) { // vdso segment can be one page or two pages.
-      rc2 = mtcp_sys_mremap(vdsoStart, 1*PAGESIZE, 1*PAGESIZE,
-		            MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
+    setupTempRegions();
+    Area area;
+    int mapsfd = mtcp_sys_open2("/proc/self/maps", O_RDONLY);
+    if (mapsfd < 0) {
+      MTCP_PRINTF("error opening /proc/self/maps; errno: %d\n", mtcp_sys_errno);
+      mtcp_abort();
     }
-    if (rc1 == MAP_FAILED || rc2 == MAP_FAILED) {
-      MTCP_PRINTF("mtcp_restart failed: "
-		  "gdb --mpi \"DMTCP_ROOT/bin/mtcp_restart\" to debug.\n");
+
+    while (mtcp_readmapsline(mapsfd, &area)) {
+      void *rc = 0x0;
+      if (mtcp_strcmp(area.name, "[vvar]") == 0) {
+        rc = mtcp_sys_mremap(area.addr, area.size, area.size,
+	                     MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
+      }
+      if (mtcp_strcmp(area.name, "[vdso]") == 0) {
+        rc = mtcp_sys_mremap(area.addr, area.size, area.size,
+                             MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
+      }
+      if (rc == MAP_FAILED) {
+        MTCP_PRINTF("mtcp_restart failed: "
+                    "gdb --mpi \"DMTCP_ROOT/bin/mtcp_restart\" to debug.\n");
+      }
     }
+    mtcp_sys_close(mapsfd);
```

Comment:
First, I think we can use one function to do them all, it doesn't make too much sense to me to separate them.

Second, why do we need to remap the vvar and vdso? Please add comments to explain.

Reply:
1. that's true.
2. noted.